### ISSUE TITLE
[Feature Request] Native support for Mayflash DolphinBar - Wiimote Emulation (Mode 4)

### ISSUE THREAD

**Issue by HeyItsLollie** [libretro/RetroArch#9732]
For reference: [Mayflash DolphinBar](http://www.mayflash.com/Products/NINTENDOWiiU/W010.html)

**TLDR:**
It'd be great to support the DolphinBar's Wiimote Emulation (Mode 4) through RetroArch itself, so that up to four Wiimotes can be used for lightgun games across all cores, _without_ having to rely on multi-mouse input.

More than willing to put a bounty on this one to help push interest in investigating this feature.

-----

The DolphinBar has been out since 2014, and is still currently [recommended by the Dolphin team](https://dolphin-emu.org/blog/2014/08/23/dolphinbar-review/) for users looking for simple Wiimote functionality — namely for its Wiimote Emulation (Mode 4 on the DolphinBar), giving it the ability to connect up to four Wiimotes directly to the Dolphin emulator via Bluetooth, complete with speaker data, IR and motion controls.

An important part of this is that Wiimote Emulation is achieved _without_ a multi-mouse setup. The DolphinBar does offer keyboard-and-mouse emulation via Mode 1 and 2, but Mode 4 is different.

I've tried looking around for word on DolphinBar support in RetroArch. From what I can tell, the current state of support appears to be:
- Lakka has had some form of DolphinBar support since 2.1, although I don't believe it supports Mode 4 specifically.
- [RetroArch's Dolphin core has DolphinBar support](https://github.com/libretro/dolphin/commit/873cab28828feef7ace9c863996b142589c62c84). This support is specifically for the Dolphin core alone, and doesn't extend to other cores.

So, I think the most sensible way of supporting DolphinBar Mode 4 would be to implement it in RetroArch itself, for a couple reasons.
1. **Supporting DolphinBar directly in RetroArch benefits all cores with lightgun support.**
This would more or less provide an input layer to push onto other cores, without having to explicitly add DolphinBar support to the cores themselves. This could also feasibly open up support for other modern-day lightguns in the future. (see: [Sinden Lightgun](http://www.sindenlightgun.com/))
2. **RetroArch could have full control over how Wiimote X/Y IR coordinates are mapped.**
Frankly, unless you're using a single monitor on Windows (or happen to be using Linux), DolphinBar's mouse emulation leaves a lot to be desired — and DolphinBar only supports emulation of a single mouse in Mode 1/2. Utilizing Mode 4 in RetroArch could help help unify the experience with Wiimote controllers, and put more control in the users' hands. For example: A user may want to recalibrate their Wiimote using four-point calibration (point at each corner of the screen).

**Comment by markwkidd** [comment]
@HeyItsLollie are you still monitoring this issue? I'll contribute $10 to start a bounty if you're still interested in it as well.

**Comment by markwkidd** [comment]
There is this C language library: https://github.com/wiiuse/wiiuse (GPL 3)

**Comment by HeyItsLollie** [comment]
> @HeyItsLollie are you still monitoring this issue? I'll contribute $10 to start a bounty if you're still interested in it as well.
I am still interested! I'm able to pitch a couple hundred towards it to get the bounty-ball rolling.
(Also oops, I hadn't seen #6466 before, glad to see that other people have brought this up)

It's obviously been a couple years since I posted this, so I'm now wondering what the best path would be — whether that's via a library like Wiiuse as you linked, Dolphin's [WiimoteReal](https://github.com/dolphin-emu/dolphin/tree/master/Source/Core/Core/HW/WiimoteReal), or something else. It's more important that any implementation supports native Wii peripherals like the Wii Sensor Bar and Wiimote, since (hopefully) that support would flow outwards to 3rd party peripherals too.

That said, getting this properly implemented would have some implications for how Retroarch handles mouse/pointer input, especially for calibration. It's a big ask. If Libretro has dedicated contributors for the controller side of things, it'd be great to hear some thoughts about what we're asking for here.

**Comment by rcl0ng** [comment]
I'd put a few dollars towards it. Has someone made a bounty? what's the link? 
https://docs.libretro.com/development/bounties/
https://www.bountysource.com/

Maybe you can muster up more support here-
https://forums.launchbox-app.com/topic/34653-guide-cheap-arcade-light-guns-with-touchemote/page/13/
or on reddit. good luck.

**Comment by rcl0ng** [comment]
https://www.bountysource.com/issues/83873923-feature-request-native-support-for-mayflash-dolphinbar-wiimote-emulation-mode-4

**Comment by stupacshakur** [comment]
This would be an amazing feature. I just got the Mayflash Dolphin Bar and its support in Dolphin is superb. Playing lightgun games in mouse mode in Retroarch does leave a lot to be desired. It works, but calibration and multiplayer would be a dream come true. I put $15 on the bounty.

**Comment by Tartifless** [comment]
Hi,
Is this feature request still open ? It would be an amazing addition for many cores (mame, fbneo, flycast, ps2, mednafen_psx (allvariants), fceumm, ...).
Many core have lightgun support and wiimotes are the less expansive way to get to it.

Having to switch between mode 2 and mode 4 is a real pain in the a...

**Comment by n33k33** [comment]
Been pulling my hair for a week trying to find solutions for using Wiimotes as Lightguns outside of Dolphin and would absolutely supporting this.

In the meantime in case it might help someone, so far I've only managed to use several Wiimotes at the same time in MAMEUI (not RA). It is for some reason the only program I've found that will detect Wiimote X / Y movements correctly. I do have to run Touchmote with motion set as X360 "Right Stick" for this tho.

I've read other people are able to play 2 players using 2 Dolphin Bars in Mode 2. I wrote to Mayflash asking them if pairing 2 Wiimotes in Mouse cursor mode would be possible at all via firmware update since programs to support several cursors do exist. Right now Mode 2 will only allow 1 Wiimote at a time.

**Comment by illando** [comment]
Bump for this request!!
It wold be a great addition also on the android version of RA where the dolphin bar works perfectly on Dolphin, but it’s not recognized in Mode2 on RA for lightgun emulation.

**Comment by rcl0ng** [comment]
This-
https://geekonarium.de/en/lichtknarre-lightgun/

**Comment by illando** [comment]
@rcl0ng Yes, but this is windows only solution.
This issue is about the possibility to have native support on RA, on the various platforms, like Dolphin has.
I don’t understand why you should use that app with dolphin when the dolphin bar is totally supported on windows.

**Comment by GeneralShakewell** [comment]
Any updates on progress? Windows recently increased driver security and severely nerfed Gunmote and lichtknarre. Could really use native mode 4 support in retroarch.

**Comment by illando** [comment]
> Any updates on progress? Windows recently increased driver security and severely nerfed Gunmote and lichtknarre. Could really use native mode 4 support in retroarch.

Yes please! The DolphinBar mode 4 is the cheaper way to play lightgun games, 1 or 2 players!
Unfortunately, its mode 1 (the only one actually working with RetroArch) supports only 1 player and relies on the OS airmouse native support.
So it works only on PC, MAC and some few LINUX distros but not on Android.
Once implemented in RetroArch it will be cross-platform as it is when used with the various version of Dolphin!
